File: benchmarks/pandas/bench_cat_ops_from_codes.py
"""
Benchmark: pd.Categorical.from_codes, reorder_categories by freq, ordered categorical on 100k elements.
Outputs JSON: {"function": "cat_ops_from_codes", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd
import numpy as np

SIZE = 100_000
WARMUP = 5
ITERATIONS = 20

categories = ["alpha", "beta", "gamma", "delta", "epsilon"]
codes = [i % len(categories) for i in range(SIZE)]
order = ["epsilon", "delta", "gamma", "beta", "alpha"]

def cat_from_codes():
    return pd.Categorical.from_codes(codes, categories=categories)

def cat_sort_by_freq(c):
    s = pd.Series(c)
    freq_order = s.value_counts().index.tolist()
    return s.astype(pd.CategoricalDtype(categories=freq_order, ordered=False))

def cat_to_ordinal(c):
    s = pd.Series(c)
    return s.astype(pd.CategoricalDtype(categories=order, ordered=True))

for _ in range(WARMUP):
    c = cat_from_codes()
    cat_sort_by_freq(c)
    cat_to_ordinal(c)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    c = cat_from_codes()
    cat_sort_by_freq(c)
    cat_to_ordinal(c)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "cat_ops_from_codes",
    "mean_ms": total_ms / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total_ms,
}))

File: benchmarks/tsb/bench_cat_ops_from_codes.ts
/**
 * Benchmark: catFromCodes, catSortByFreq, catToOrdinal on 100k elements.
 * Outputs JSON: {"function": "cat_ops_from_codes", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { catFromCodes, catSortByFreq, catToOrdinal } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 20;

const categories = ["alpha", "beta", "gamma", "delta", "epsilon"];
const codes = Array.from({ length: SIZE }, (_, i) => i % categories.length);
const order = ["epsilon", "delta", "gamma", "beta", "alpha"];

for (let i = 0; i < WARMUP; i++) {
  const cs = catFromCodes(codes, categories);
  catSortByFreq(cs);
  catToOrdinal(cs, order);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  const cs = catFromCodes(codes, categories);
  catSortByFreq(cs);
  catToOrdinal(cs, order);
  times.push(performance.now() - t0);
}

const total = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "cat_ops_from_codes",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
